<?xml version="1.0"?>
<robot name="tricycle_robot"> 

  <!-- Base -->
  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.4 0.3 0.2"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="blue">
        <color rgba="0 0 1 1"/>
      </material>
    </visual>
    <collision>
        <geometry>
            <box size="0.4 0.3 0.2"/>
        </geometry>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>
    <inertial>
        <mass value="1"/>
        <inertia ixx="0.01083" ixy="0" ixz="0" iyy="0.01667" iyz="0" izz="0.02083"/>
    </inertial>
  </link>

  <!-- Rear Left Wheel -->
  <link name="left_wheel">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="pink">
        <color rgba="1 0 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.04" radius="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>
    <inertial>
        <mass value="0.1"/>
        <inertia ixx="7.58e-5" ixy="0" ixz="0" iyy="7.58e-5" iyz="0" izz="0.000125"/>
    </inertial>
  </link>

  <!-- Rear Right Wheel  -->
  <link name="right_wheel">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="pink">
        <color rgba="1 0 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.04" radius="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>
    <inertial>
        <mass value="0.1"/>
        <inertia ixx="7.58e-5" ixy="0" ixz="0" iyy="7.58e-5" iyz="0" izz="0.000125"/>
    </inertial>
  </link>

  <!-- FRONT STEERING WHEEL  -->

  <link name="front_fork_link">
    <visual>
      <geometry>
        <cylinder length="0.15" radius="0.01"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="gray">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
        <geometry>
            <cylinder length="0.15" radius="0.01"/>
        </geometry>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>
    <inertial>
        <mass value="0.2"/>
        <inertia ixx="2.5e-4" ixy="0" ixz="0" iyy="2.5e-4" iyz="0" izz="2.5e-4"/>
    </inertial>
  </link>

  <!-- Front wheel link -->
  <link name="front_wheel">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="pink">
        <color rgba="1 0 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.04" radius="0.05"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>
    <inertial>
        <mass value="0.1"/>
        <inertia ixx="7.58e-5" ixy="0" ixz="0" iyy="7.58e-5" iyz="0" izz="0.000125"/>
    </inertial>
  </link>

  <!-- CAMERA LINK -->

  <link name="camera_link">
    <visual>
      <geometry>
        <box size="0.04 0.04 0.04"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="white">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.04 0.04 0.04"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="6.58e-5" ixy="0" ixz="0"
             iyy="6.58e-5" iyz="0" izz="0.000125"/>
    </inertial>
  </link>



  <!-- ========== JOINTS ========== -->

  <joint name="base_to_left_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="left_wheel"/>
    <origin xyz="-0.13 0.15 -0.1" rpy="1.57 0 0"/>
    <axis xyz="0 0 1"/>
    <dynamics damping="0.0" friction="0.1"/>
  </joint>

  <joint name="base_to_right_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="right_wheel"/>
    <origin xyz="-0.13 -0.15 -0.1" rpy="1.57 0 0"/>
    <axis xyz="0 0 1"/>
    <dynamics damping="0.0" friction="0.1"/>
  </joint>


  <!-- Steering joint: base_link -> front_fork_link -->
  <joint name="front_steer_joint" type="revolute">
    <parent link="base_link"/>
    <child link="front_fork_link"/>
    <origin xyz="0.2 0 -0.02" rpy="0 0 0"/>
    <!-- steering around vertical axis -->
    <axis xyz="0 0 1"/>
    <limit lower="-0.4" upper="0.4" effort="10" velocity="1.0"/>
  </joint>

  <!-- Wheel spin joint: front_fork_link -> front_wheel -->
  <joint name="front_wheel_spin_joint" type="continuous">
    <parent link="front_fork_link"/>
    <child link="front_wheel"/>
    <!-- slightly below fork center -->
    <origin xyz="0 0 -0.08" rpy="1.57 0 0"/>
    <!-- wheel spins around its axle (Y) -->
    <axis xyz="0 0 1"/>
  </joint>

  <joint name="base_to_camera" type="fixed">
    <parent link="base_link"/>
    <child link="camera_link"/>
    <origin xyz="0.15 0 0.1" rpy="0 0 0"/>
  </joint>

  <!-- Gazebo Materials       -->

  <gazebo reference="base_link">
    <material>Gazebo/Blue</material>
  </gazebo>

  <gazebo reference="left_wheel">
    <material>Gazebo/Pink</material>
  </gazebo>

  <gazebo reference="right_wheel">
    <material>Gazebo/Pink</material>
  </gazebo>

  <gazebo reference="front_wheel">
    <material>Gazebo/Gray</material>
    <mu1 value="0.8"/>
    <mu2 value="0.8"/>
  </gazebo>

  <gazebo reference="camera_link">
    <material>Gazebo/White</material>
  </gazebo>


  <!-- Gazebo ROS Control Plugin -->
  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so"/>
  </gazebo>

  <!-- Differential Drive Plugin (rear-wheel drive) -->
  <gazebo>
    <plugin name="diff_drive_controller" filename="libgazebo_ros_diff_drive.so">
      <ros>
        <remapping>cmd_vel:=/cmd_vel</remapping>
      </ros>

      <!-- Wheel Information -->
      <left_joint>base_to_left_wheel</left_joint>
      <right_joint>base_to_right_wheel</right_joint>
      <wheel_separation>0.3</wheel_separation>
      <wheel_diameter>0.1</wheel_diameter>

 
      <max_wheel_torque>200</max_wheel_torque>
      <max_wheel_acceleration>10.0</max_wheel_acceleration>


      <odometry_frame>odom</odometry_frame>
      <robot_base_frame>base_link</robot_base_frame>

      <publish_odom>true</publish_odom>
      <publish_odom_tf>true</publish_odom_tf>
      <publish_wheel_tf>true</publish_wheel_tf>
    </plugin>
  </gazebo>

    <!-- CAMERA PLUGIN -->
    <gazebo reference="camera_link">
    <sensor name="front_camera" type="camera">
      <pose>0 0 0 0 0 0</pose>
      <camera>
        <horizontal_fov>1.3962634</horizontal_fov>
        <image>
          <width>640</width>
          <height>480</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.05</near>
          <far>0.5</far>
        </clip>
      </camera>

      <always_on>true</always_on>
      <update_rate>15.0</update_rate>
      <visualize>true</visualize>

      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <ros>
          <namespace>/</namespace>
          <remapping>image_raw:=/camera/image_raw</remapping>
          <remapping>camera_info:=/camera/camera_info</remapping>
        </ros>
        <frame_name>camera_link</frame_name>
        <camera_name>front_camera</camera_name>
      </plugin>
    </sensor>
  </gazebo>



</robot>
